New calling convention preserve_none

weiguozhi opened the pull request:
The new experimental calling convention preserve_none is the opposite side of existing preserve_all. It tries to preserve as few general registers as possible. So all general registers are caller saved registers. It can also uses more general registers to pass arguments. This attribute doesn't impact floating-point registers. Floating-point registers still follow the c calling convention.

Currently preserve_none is supported on X86-64 only. It changes the c calling convention in following fields:
  
  * RSP and RBP are the only preserved general registers, all other general registers are caller saved registers.
  * We can use [RDI, RSI, RDX, RCX, R8, R9, R11, R12, R13, R14, R15, RAX] to pass arguments.

It can improve the performance of hot tailcall chain, because many callee saved registers' save/restore instructions can be removed if the tail functions are using preserve_none. In my experiment in protocol buffer, the parsing functions are improved by 3% to 10%.

weiguozhi committed:
New calling convention preserve_none

The new calling convention preserve_none is the opposite side of
existing preserve_all. It tries to preserve as few general registers as
possible. So all general registers are caller saved registers. It can
also uses more general registers to pass arguments. This attribute
doesn't impact floating-point registers. Floating-point registers still
follow the c calling convention.

Currently preserve_none is supported on X86-64 only.

pinskia commented on llvm/docs/LangRef.rst:
```diff
@@ -416,6 +416,12 @@ added in the future:
+"``preserve_nonecc``" - The `PreserveNone` calling convention
+    This calling convention doesn't preserve any general registers. So all
+    general registers are caller saved registers. It also uses all general
+    registers to pass arguments. This attribute doesn't impact floating-point
```
I suspect this should say any `non-general purpose registers (e.g. floating point registers, on x86 XMMs/YMMs)`. Rather than `floating-point registers`.
Also isn't this just a hack to increase the number of registers used to pass arguments? If so there has to be a better way of doing this.  Maybe a non-exposed attribute which is used only for non-exposed functions?
e.g. on x86 (not 64bit), regparm could be used internally there.

weiguozhi replied:
> I suspect this should say any `non-general purpose registers (e.g. floating point registers, on x86 XMMs/YMMs)`. Rather than `floating-point registers`. 

Thanks for the correction! Will change it.

> Also isn't this just a hack to increase the number of registers used to pass arguments? If so there has to be a better way of doing this. Maybe a non-exposed attribute which is used only for non-exposed functions? e.g. on x86 (not 64bit), regparm could be used internally there.

It's not a hack, it's a natural extension. Because we don't preserve general registers, we can use all of them to pass arguments without the extra cost to save/restore those registers, they are clobbered by the function call anyway.

KanRobert commented on llvm/lib/Target/X86/X86CallingConv.td:
```diff
@@ -1056,6 +1056,22 @@ def CC_Intel_OCL_BI : CallingConv<[
+def CC_X86_64_Preserve_None : CallingConv<[
+  // We don't preserve general registers, so all of them can be used to pass
+  // arguments except
+  //   - RBP  frame pointer
```
It does not mention R16-R31

weiguozhi replied:
Added a comment to mention that R16-R31 are not used to pass arguments because they are not universally available.

weiguozhi committed:
Update the comment and document.

aeubanks commented:
the clang changes should be split into a followup patch

erichkeane commented on clang/include/clang/Basic/Attr.td:
```diff
@@ -2868,6 +2868,11 @@ def M68kRTD: DeclOrTypeAttr {
 }
 
+def PreserveNone : DeclOrTypeAttr {
+  let Spellings = [Clang<"preserve_none">];
```
This should have a subjectlist set, since this is on functions, it should probably have a function subject.  Since it is target specific, it should ALSO be a target specific attribute.

weiguozhi replied:
I tried to add
```
  let Subjects = SubjectList<[Function]>;
```
But it doesn't work for the code
```
  void __attribute__((preserve_none)) boo(void *ptr) {
  }

  void (__attribute__((preserve_none)) *pboo1)(void *) = boo;
```
I got the error message
```
'preserve_none' attribute only applies to functions
```

erichkeane replied:
The second example there is a function pointer, not a function you're trying to apply it to.  If you want it to work on Function Pointers as well, you need to use `FunctionLike`.

weiguozhi replied:
FunctionLike works for me. Thanks!

erichkeane reviewed (commented):
CFE changes aren't bad, just a few comments on how the attribute is defined in attr.td

jyknight commented on llvm/lib/Target/X86/X86CallingConv.td:
```diff
@@ -1056,6 +1056,23 @@ def CC_Intel_OCL_BI : CallingConv<[
+                                 R11, R12, R13, R14, R15, RAX]>>,
+
+  // Otherwise it's the same as the regular C calling convention.
+  CCDelegateTo<CC_X86_64_C>
```
This delegation seems questionable -- what about the interaction with the swift attributes which use dedicated registers in CC_X86_64_C, which you're now also using for normal parameters?

weiguozhi replied:
I have never thought of swift attributes. How can I disable swift attributes when using preserve_none calling convention?

weiguozhi replied:
I added code to check if there are any swift attributes used with preserve_none at the same time, and report error if it is detected.

weiguozhi replied:
@jyknight could you help to review the code that checks the combination of swift attributes and preserve_none in https://github.com/llvm/llvm-project/pull/76868/commits/dc1bc55634045adb5b0ab465b16dcf73f1e77144

KanRobert commented on llvm/lib/Target/X86/X86CallingConv.td:
```diff
@@ -1056,6 +1056,23 @@ def CC_Intel_OCL_BI : CallingConv<[
+  //   - RBP        frame pointer
+  //   - R10        'nest' parameter
+  //   - RBX        base pointer
+  //   - R16 - R31  these are not available everywhere
```
I won't expect this from the name `preserve_none`. R16-R31 should be used to pas arguments when they're available.

jyknight replied:
They should not be used to pass arguments conditionally based on the current subtarget, since that creates two incompatible calling conventions.

There's no reason "preserve none" has to be read to imply "uses all possible registers to pass arguments.", so I don't see an issue with leaving it like it is.

weiguozhi committed:
New calling convention preserve_none

Specify it has a FunctionLike subject.
Specify it as a target specific attribute.

weiguozhi committed:
New calling convention preserve_none

Report error when swift attribute is used with preserve_none at the same
time.

rnk reviewed (approved):
This looks complete to the best of my knowledge, but I know others have specific concerns, like James identified the swift attribute conflict. To other reviewers, do you have any outstanding concerns to address?

I think it would be nice to quickly generalize this to AArch64 if possible, since all the same use cases apply there.

david-xl commented on llvm/test/CodeGen/X86/preserve_nonecc64-ret-double.ll:
```diff
@@ -0,0 +1,85 @@
+define preserve_nonecc double @preserve_nonecc1() nounwind {
+entry:
+;ALL-LABEL:   preserve_nonecc1
+;ALL-NOT:     movaps %xmm1
```
check no savings of GSRs too?

david-xl commented on llvm/test/CodeGen/X86/preserve_nonecc64-ret-double.ll:
```diff
@@ -0,0 +1,85 @@
+declare preserve_nonecc double @bar_double(i64, i64)
+define void @preserve_nonecc2() nounwind {
+entry:
+;ALL-LABEL: preserve_nonecc2
```
Those xmm registers will be preserved/restored around the call even without the attribute. Perhaps check the the GSRs ?

david-xl reviewed (commented):
Can you increase the test coverage:

1) preserve_none caller TAIL-calls preserve_none callee -- the tail call is preserved
2) regular function TAIL-calls preserve_none callee -- the tail call should be disabled and all CSRs should be saved/restored around the call.
3) preserve-none caller calls preserve_none callee -- no registers are saved/restored around the call
4) preserve_none caller calls preserve_none callee with long argument list -- CSRs used for arg passing.

weiguozhi committed:
New calling convention preserve_none

Add more test cases.

david-xl commented on llvm/test/CodeGen/X86/preserve_nonecc_call.ll:
```diff
@@ -0,0 +1,131 @@
+}
+
+; Preserve_none function can use more registers to pass parameters.
+define preserve_nonecc i64 @callee_with_many_param(i64 %a1, i64 %a2, i64 %a3, i64 %a4, i64 %a5, i64 %a6, i64 %a7, i64 %a8, i64 %a9, i64 %a10, i64 %a11, i64 %a12) {
```
To test whether the callee/caller register (parameters) sequence matches, it is probably better avoiding using the adding which are subject to re-association. How about do the following
g1 = a1;
g2 = a2;
...

where g1, g2, etc are global variables.

david-xl replied:
or better yet, pass the arguments in order to (tail)call to another preserve_none callee.

weiguozhi committed:
New calling convention preserve_none

Enhance a test case.

weiguozhi committed:
New calling convention preserve_none

Modify the test case to avoid possible future re-association impact.

brandtbucher commented:
Thanks for working on this! It's really appreciated.

Two follow-up questions:
- Is AArch64 support planned soon?
- Am I correct in understanding that this missed the LLVM 18 cut, and won't until LLVM 19 in mid-August-ish?

calebh commented:
We have another use case for this calling convention - micropatching. See my coworker's blog post on how this would work: https://www.philipzucker.com/permutation_compile/

Is there a reason why r10 isn't included in the list of registers for x64?

Having support for additional architectures (ARM, PPC) would be great for our use case.

weiguozhi commented:
> We have another use case for this calling convention - micropatching. See my coworker's blog post on how this would work: https://www.philipzucker.com/permutation_compile/
> 
I'm happy to see it's useful to you.

> Is there a reason why r10 isn't included in the list of registers for x64?
> 
In the standard C calling convention R10 is used as nest parameter, I don't know its exact use case (maybe for nest function?). Excluding R10 will avoid problem in those cases.

> Having support for additional architectures (ARM, PPC) would be great for our use case.
We are also interested in AArch64 support, but currently it's not in our high priority.

brandtbucher commented:
After playing around with this a bit, I'm wondering if the registers used for argument-passing should be reordered.

The motivating use-case for this calling convention is to pin registers across chained tail calls, but much of the pinning effect is lost if any "normal" C calls are made. This is because the first 7 parameters are all being pinned in volatile registers, so any call to a non-`preserve_none` function results in a cascade of shuffles and spills. You can see a simple example of this here: https://godbolt.org/z/rzsG9fb4W

If I understand correctly, we can reduce this effect by choosing registers that are normally callee-saved (`r12`, `r13`, `r14`, and `r15`) *before* using normally-clobbered registers (`rdi`, `rsi`, `rdx`, `rcx`, `r8`, `r9`, `r11`, and `rax`). This is what the similar GHC calling convention does.

What do you think, @weiguozhi?

david-xl commented:
> After playing around with this a bit, I'm wondering if the registers used for argument-passing should be reordered.
> 
> The motivating use-case for this calling convention is to pin registers across chained tail calls, but much of the pinning effect is lost if any other calls are made in the body of the function. This is because the first 7 parameters are all being pinned in volatile registers, so any call results in a cascade of shuffles and spills. You can see a simple example of this here: https://godbolt.org/z/rzsG9fb4W
> 
> If I understand correctly, we can reduce this effect by choosing registers that are normally callee-saved (`r12`, `r13`, `r14`, and `r15`) _before_ using normally-clobbered registers (`rdi`, `rsi`, `rdx`, `rcx`, `r8`, `r9`, `r11`, and `rax`). This is what the similar GHC calling convention does.
> 
> What do you think, @weiguozhi?

Makes sense. @weiguozhi can you also help measure the performance impact with the order reverse change with preserve_none?  I wonder if a new calling convention is needed or just bake this into preserve_none.

brandtbucher commented:
Draft PR at https://github.com/llvm/llvm-project/pull/88333.

Keno commented:
I was wondering what the motivation was to still preserve `rbp`? I guess `-fomit-frame-pointers` has fallen a bit out of favor, but I'd have expected `preserve_none`, to preserve, well, none.
